In [2]:
import pandas as pd
import numpy as np
import re as re


In [4]:
#df reading

df_ventas = pd.read_csv("D:/Desktop/items_ordered_2years.csv", sep=",")

In [3]:
df_ventas

num_order                           item_id  \
0       ce30c2f02458457e3c7b563a636ae2a1  0916c05c5c3f65f59d813a78ac35c8d2   
1       ce30c2f02458457e3c7b563a636ae2a1  ff323b39ae36843396d2e53ce549fb10   
2       ce30c2f02458457e3c7b563a636ae2a1  199916dffc95259f4d2daab6664ca9c0   
3       83e75d608f11c8163599806420903ab9  8ca334ec2493501139327ce0165a1a84   
4       9b687270d8e7eed9717022af5961a190  ce0a1683c6b1a0248b330344ec592ddf   
...                                  ...                               ...   
930955  479ed436c26a511c4c3356e679bccbc9  37d700d54a6acc0ff49392cce34b6427   
930956  479ed436c26a511c4c3356e679bccbc9  05fd04261f1c951b07f2ac3a67254d81   
930957  bf923aa16e0608a5ed95347a9f411dce  8d8600ea4e451330f8614ba08ad4f047   
930958  bf923aa16e0608a5ed95347a9f411dce  b74b2e1a096c656a247142e24eefb738   
930959  bf923aa16e0608a5ed95347a9f411dce  fd0c1bb9a90dce3f3b2099990f019262   

                 created_at  product_id  qty_ordered  base_cost  price  \
0       2018-11-06 16:52:13       86434          1.0     9.3750  15.38   
1       2018-11-06 16:52:13       87652          1.0     7.0000  15.41   
2       2018-11-06 16:52:13        2785          1.0     9.1587  12.23   
3       2018-12-17 12:26:54       13000          5.0    13.7936  15.03   
4       2017-01-12 14:19:03       41947          1.0     5.5924   7.00   
...                     ...         ...          ...        ...    ...   
930955  2018-12-30 23:51:57        1868          1.0     4.0700   4.79   
930956  2018-12-30 23:51:57       23539          1.0     4.0700   5.35   
930957  2018-12-30 23:57:37       92793          1.0    33.7360  32.93   
930958  2018-12-30 23:57:37       93036          1.0     0.0000  16.32   
930959  2018-12-30 23:57:37       11850          1.0    10.4414  12.53   

        discount_percent                       customer_id  \
0                    8.0  eccbc87e4b5ce2fe28308fd9f2a7baf3   
1                    8.0  eccbc87e4b5ce2fe28308fd9f2a7baf3   
2                    8.0  eccbc87e4b5ce2fe28308fd9f2a7baf3   
3                   50.0  eccbc87e4b5ce2fe28308fd9f2a7baf3   
4                    5.0  1679091c5a880faf6fb5e6087eb1b2dc   
...                  ...                               ...   
930955              10.0  d9cd4886b615a36059f72a74d7019113   
930956              10.0  d9cd4886b615a36059f72a74d7019113   
930957              10.0  d73f27d3ab96df71f0fdc47511f326bc   
930958              10.0  d73f27d3ab96df71f0fdc47511f326bc   
930959              10.0  d73f27d3ab96df71f0fdc47511f326bc   

                         city zipcode  
0       TARAZONA DE LA MANCHA   02100  
1       TARAZONA DE LA MANCHA   02100  
2       TARAZONA DE LA MANCHA   02100  
3       TARAZONA DE LA MANCHA   02100  
4                    Alboraya   46120  
...                       ...     ...  
930955                  Ubeda   23400  
930956                  Ubeda   23400  
930957                    Vic   08500  
930958                    Vic   08500  
930959                    Vic   08500  

[930960 rows x 11 columns]

In [4]:
#Duplicates checking
df_ventas = df_ventas.drop_duplicates()

df_ventas

num_order                           item_id  \
0       ce30c2f02458457e3c7b563a636ae2a1  0916c05c5c3f65f59d813a78ac35c8d2   
1       ce30c2f02458457e3c7b563a636ae2a1  ff323b39ae36843396d2e53ce549fb10   
2       ce30c2f02458457e3c7b563a636ae2a1  199916dffc95259f4d2daab6664ca9c0   
3       83e75d608f11c8163599806420903ab9  8ca334ec2493501139327ce0165a1a84   
4       9b687270d8e7eed9717022af5961a190  ce0a1683c6b1a0248b330344ec592ddf   
...                                  ...                               ...   
930955  479ed436c26a511c4c3356e679bccbc9  37d700d54a6acc0ff49392cce34b6427   
930956  479ed436c26a511c4c3356e679bccbc9  05fd04261f1c951b07f2ac3a67254d81   
930957  bf923aa16e0608a5ed95347a9f411dce  8d8600ea4e451330f8614ba08ad4f047   
930958  bf923aa16e0608a5ed95347a9f411dce  b74b2e1a096c656a247142e24eefb738   
930959  bf923aa16e0608a5ed95347a9f411dce  fd0c1bb9a90dce3f3b2099990f019262   

                 created_at  product_id  qty_ordered  base_cost  price  \
0       2018-11-06 16:52:13       86434          1.0     9.3750  15.38   
1       2018-11-06 16:52:13       87652          1.0     7.0000  15.41   
2       2018-11-06 16:52:13        2785          1.0     9.1587  12.23   
3       2018-12-17 12:26:54       13000          5.0    13.7936  15.03   
4       2017-01-12 14:19:03       41947          1.0     5.5924   7.00   
...                     ...         ...          ...        ...    ...   
930955  2018-12-30 23:51:57        1868          1.0     4.0700   4.79   
930956  2018-12-30 23:51:57       23539          1.0     4.0700   5.35   
930957  2018-12-30 23:57:37       92793          1.0    33.7360  32.93   
930958  2018-12-30 23:57:37       93036          1.0     0.0000  16.32   
930959  2018-12-30 23:57:37       11850          1.0    10.4414  12.53   

        discount_percent                       customer_id  \
0                    8.0  eccbc87e4b5ce2fe28308fd9f2a7baf3   
1                    8.0  eccbc87e4b5ce2fe28308fd9f2a7baf3   
2                    8.0  eccbc87e4b5ce2fe28308fd9f2a7baf3   
3                   50.0  eccbc87e4b5ce2fe28308fd9f2a7baf3   
4                    5.0  1679091c5a880faf6fb5e6087eb1b2dc   
...                  ...                               ...   
930955              10.0  d9cd4886b615a36059f72a74d7019113   
930956              10.0  d9cd4886b615a36059f72a74d7019113   
930957              10.0  d73f27d3ab96df71f0fdc47511f326bc   
930958              10.0  d73f27d3ab96df71f0fdc47511f326bc   
930959              10.0  d73f27d3ab96df71f0fdc47511f326bc   

                         city zipcode  
0       TARAZONA DE LA MANCHA   02100  
1       TARAZONA DE LA MANCHA   02100  
2       TARAZONA DE LA MANCHA   02100  
3       TARAZONA DE LA MANCHA   02100  
4                    Alboraya   46120  
...                       ...     ...  
930955                  Ubeda   23400  
930956                  Ubeda   23400  
930957                    Vic   08500  
930958                    Vic   08500  
930959                    Vic   08500  

[915165 rows x 11 columns]

In [5]:
df_ventas.nunique()

num_order           278045
item_id             906317
created_at          277896
product_id           26396
qty_ordered             57
base_cost            52202
price                 5028
discount_percent        45
customer_id         118799
city                 21012
zipcode              11132
dtype: int64

In [6]:
df_ventas.isna().sum()

num_order              0
item_id                0
created_at             0
product_id             0
qty_ordered            0
base_cost           2358
price                  0
discount_percent       0
customer_id            0
city                2910
zipcode             2910
dtype: int64

### item_id

In [7]:
df_ventas["item_id"].nunique()

906317

In [8]:
item_nu = df_ventas.groupby(
    "item_id", as_index=False
).size()

mascara = item_nu.iloc[:,1] > 1

item_repetidos = item_nu[mascara]

item_repetidos

item_id  size
111     000802dde903f8282df756d2aab8d024     2
135     000a8c9981c4dc672abd207bdc1b85f7     2
236     001183d1253fa3d736eda6c92c826360     2
375     001b62e0e6c6fcf7fd771422bc91cb12     2
385     001bd90aa12b86fc11c4a416b3495cd5     2
...                                  ...   ...
906117  fff249858b2d160946da7b8613aa15da     2
906126  fff28e48c4498626a0add97bb3c9cb55     2
906144  fff3cf8f9ffdc3188bfb0b0546b37245     2
906292  fffe4633ea7afca58a2a1e68f2963f6b     2
906306  ffff1e771c21fcbfed80ea59dcb7bf5d     2

[8848 rows x 2 columns]

In [9]:
mascara2 = df_ventas["item_id"].isin(
    item_repetidos["item_id"].tolist()
)

df_ventas_repetido = df_ventas[mascara2].sort_values("item_id")

df_ventas_repetido

num_order                           item_id  \
649241  8051708bf244ba8518b5e5b61a74486b  000802dde903f8282df756d2aab8d024   
649259  8051708bf244ba8518b5e5b61a74486b  000802dde903f8282df756d2aab8d024   
447200  6d19dedf8710788ea7d774c3fc977a9b  000a8c9981c4dc672abd207bdc1b85f7   
447182  6d19dedf8710788ea7d774c3fc977a9b  000a8c9981c4dc672abd207bdc1b85f7   
766220  27df73640a5bbe494d1db3e3bf16b906  001183d1253fa3d736eda6c92c826360   
...                                  ...                               ...   
370897  bfffc8aa56b415ede4538a4a022c771a  fff3cf8f9ffdc3188bfb0b0546b37245   
91194   d69c15626b6406bbcc1384407e75a075  fffe4633ea7afca58a2a1e68f2963f6b   
91255   d69c15626b6406bbcc1384407e75a075  fffe4633ea7afca58a2a1e68f2963f6b   
803228  4f8b96d2253a9d4a6ad4672b4e228c48  ffff1e771c21fcbfed80ea59dcb7bf5d   
803223  4f8b96d2253a9d4a6ad4672b4e228c48  ffff1e771c21fcbfed80ea59dcb7bf5d   

                 created_at  product_id  qty_ordered  base_cost  price  \
649241  2018-12-17 23:32:08        7647          1.0     2.5438   3.00   
649259  2018-12-17 23:32:08        7647          1.0     2.5438   3.00   
447200  2018-09-02 16:35:36        3076          2.0     6.3986  12.71   
447182  2018-09-02 16:35:36        3076          2.0     6.3986  12.71   
766220  2018-08-07 09:29:02        2021          1.0    10.2800  15.87   
...                     ...         ...          ...        ...    ...   
370897  2017-09-18 18:50:26        2503          2.0     5.0720   6.57   
91194   2018-10-31 09:23:06        3290          1.0     4.1466   5.59   
91255   2018-10-31 09:23:06        3290          1.0     4.1466   5.59   
803228  2018-04-23 20:52:23       68375          1.0     5.2100   6.95   
803223  2018-04-23 20:52:23       68375          1.0     5.2100   6.95   

        discount_percent                       customer_id  \
649241               5.0  c2d3e675fedd5555ee14b6441f0a8568   
649259               5.0  c2d3e675fedd5555ee14b6441f0a8568   
447200              15.0  9bd04a4cc151456b49c699624bb83084   
447182              15.0  9bd04a4cc151456b49c699624bb83084   
766220               5.0  0a6f053fb763ab1d4b6bef36ce683e53   
...                  ...                               ...   
370897               7.0  b6b909c1193aba68c413cc1442743ac7   
91194                8.0  91ae48805e8e96055d454f043054059b   
91255                8.0  91ae48805e8e96055d454f043054059b   
803228               5.0  a99855ff1ad46db840df3a48bf075d17   
803223               5.0  a99855ff1ad46db840df3a48bf075d17   

                       city zipcode  
649241   CORRALES DE BUELNA   39400  
649259   Corrales de buelna   39400  
447200             Granada    18009  
447182              Granada   18009  
766220            barcelona   08025  
...                     ...     ...  
370897               MADRID   28002  
91194   Alameda de la Sagra   45240  
91255               PANTOJA   45290  
803228               CANGAS   36940  
803223                 POIO   36163  

[17696 rows x 11 columns]

In [10]:
# tolower
# dropduplicates por item_id

In [11]:
df_ventas_sin_pueblos = df_ventas_repetido.drop(["zipcode", "city"], inplace = True, axis=1)

### Primeras conclusiones


- El dataframe tiene valores duplicados (150k aprox)
    - No será necesario depurar la variable, al ser simplemente identificadora
- item_id es un identificador único para cada producto vendido, que se puede utilizar como index. 
    - Contiene valores duplicados (8848 valores unicos repetidos dos veces cada uno, 17696 en total)
    - Esta duplicación viene dada por la ciudad y principalmente el zipcode, lo cual es extraño, al ser el item_id valores únicos
-Propuesta:
    - Depurando la columna city como se propuso soluciona parte de los problemas de city
    - Se puede preguntar a Carlos
    - Añadiendo columnas de provincia y ccaa se soluciona. En modelos, esta columna no va a tener valor, y en visualización por pueblos me parece complicado la verdad, más teniendo en cuenta que habrá pueblos con 1 producto vendido, por ejemplo.


### product_id

In [12]:
df_ventas["longitud_product_id"] = df_ventas["product_id"].astype("string").apply(len)


C:\Users\Daniel\AppData\Local\Temp\ipykernel_14124\2179094988.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_ventas["longitud_product_id"] = df_ventas["product_id"].astype("string").apply(len)


In [13]:
df_ventas

num_order                           item_id  \
0       ce30c2f02458457e3c7b563a636ae2a1  0916c05c5c3f65f59d813a78ac35c8d2   
1       ce30c2f02458457e3c7b563a636ae2a1  ff323b39ae36843396d2e53ce549fb10   
2       ce30c2f02458457e3c7b563a636ae2a1  199916dffc95259f4d2daab6664ca9c0   
3       83e75d608f11c8163599806420903ab9  8ca334ec2493501139327ce0165a1a84   
4       9b687270d8e7eed9717022af5961a190  ce0a1683c6b1a0248b330344ec592ddf   
...                                  ...                               ...   
930955  479ed436c26a511c4c3356e679bccbc9  37d700d54a6acc0ff49392cce34b6427   
930956  479ed436c26a511c4c3356e679bccbc9  05fd04261f1c951b07f2ac3a67254d81   
930957  bf923aa16e0608a5ed95347a9f411dce  8d8600ea4e451330f8614ba08ad4f047   
930958  bf923aa16e0608a5ed95347a9f411dce  b74b2e1a096c656a247142e24eefb738   
930959  bf923aa16e0608a5ed95347a9f411dce  fd0c1bb9a90dce3f3b2099990f019262   

                 created_at  product_id  qty_ordered  base_cost  price  \
0       2018-11-06 16:52:13       86434          1.0     9.3750  15.38   
1       2018-11-06 16:52:13       87652          1.0     7.0000  15.41   
2       2018-11-06 16:52:13        2785          1.0     9.1587  12.23   
3       2018-12-17 12:26:54       13000          5.0    13.7936  15.03   
4       2017-01-12 14:19:03       41947          1.0     5.5924   7.00   
...                     ...         ...          ...        ...    ...   
930955  2018-12-30 23:51:57        1868          1.0     4.0700   4.79   
930956  2018-12-30 23:51:57       23539          1.0     4.0700   5.35   
930957  2018-12-30 23:57:37       92793          1.0    33.7360  32.93   
930958  2018-12-30 23:57:37       93036          1.0     0.0000  16.32   
930959  2018-12-30 23:57:37       11850          1.0    10.4414  12.53   

        discount_percent                       customer_id  \
0                    8.0  eccbc87e4b5ce2fe28308fd9f2a7baf3   
1                    8.0  eccbc87e4b5ce2fe28308fd9f2a7baf3   
2                    8.0  eccbc87e4b5ce2fe28308fd9f2a7baf3   
3                   50.0  eccbc87e4b5ce2fe28308fd9f2a7baf3   
4                    5.0  1679091c5a880faf6fb5e6087eb1b2dc   
...                  ...                               ...   
930955              10.0  d9cd4886b615a36059f72a74d7019113   
930956              10.0  d9cd4886b615a36059f72a74d7019113   
930957              10.0  d73f27d3ab96df71f0fdc47511f326bc   
930958              10.0  d73f27d3ab96df71f0fdc47511f326bc   
930959              10.0  d73f27d3ab96df71f0fdc47511f326bc   

                         city zipcode  longitud_product_id  
0       TARAZONA DE LA MANCHA   02100                    5  
1       TARAZONA DE LA MANCHA   02100                    5  
2       TARAZONA DE LA MANCHA   02100                    4  
3       TARAZONA DE LA MANCHA   02100                    5  
4                    Alboraya   46120                    5  
...                       ...     ...                  ...  
930955                  Ubeda   23400                    4  
930956                  Ubeda   23400                    5  
930957                    Vic   08500                    5  
930958                    Vic   08500                    5  
930959                    Vic   08500                    5  

[915165 rows x 12 columns]

In [14]:

df_ventas["longitud_product_id"].unique().tolist()

[5, 4, 3, 2]

In [15]:
mascara_len = df_ventas["longitud_product_id"] < 3

In [16]:
df_ventas[mascara_len]

num_order                           item_id  \
5010    1301d57a9c46a0da4c13b33812d9513f  1ab5c7cd9926df4a281e4e35e7e7f998   
5018    8068cabfe1f4d630752965763e510ef6  8d5a6331e647f2a8e120097e8032f963   
5019    c9e26c814bcf2c8f8a8571b5871e254e  f1e934c774d8877895018d1a0df3e6c8   
5029    d422f0a7bb82094088c7dbafd1d1d934  c29fedf4f37ff04b4f22e779a4fa80a2   
7051    68944efaf14ccf437dd67567ee023805  3df6efcd2dfb75cc8b6221a54e302959   
...                                  ...                               ...   
803888  1fe7ba3dd00af9e305f416166bb04bcb  f44e02dd9db9088d82141e08f7c38dc0   
837696  330fe1846e7cbc30de8f7d50bc26b198  2d2b53af7a98f108a24001a54212834e   
837702  a277f0223fa418dffdbc5706fef83a9a  da23aca79bbb9a79c29da3184c0b92fb   
887179  9d83b74408e32be1d38ec79806e8913c  f9bec61be92af8aef51c745120bcd190   
903926  931095bb832a07d1844a186cdea775d2  809c40377f7cb8f4f79917ce1fd8486c   

                 created_at  product_id  qty_ordered  base_cost  price  \
5010    2017-01-22 22:52:15          91          1.0    14.8915  15.66   
5018    2017-10-01 21:53:25          91          1.0    11.8802  15.66   
5019    2018-03-21 22:20:27          91          2.0    13.8845  14.55   
5029    2018-10-01 21:46:59          91          1.0    12.6140  15.76   
7051    2018-10-05 15:17:36          91          2.0    12.6140  15.76   
...                     ...         ...          ...        ...    ...   
803888  2018-10-19 17:22:10          95          1.0    18.2200  22.78   
837696  2018-12-27 11:24:18          91          1.0    11.0897  14.57   
837702  2018-12-27 11:34:32          91          1.0    11.0897  14.57   
887179  2018-11-02 10:41:57          91          1.0    11.4245  15.41   
903926  2018-10-29 19:12:22          98          1.0     9.2566  12.67   

        discount_percent                       customer_id             city  \
5010                 7.0  75429d136f65d2d6168b9b6c5f6ec951  HINOJAR DEL REY   
5018                15.0  75429d136f65d2d6168b9b6c5f6ec951  HINOJAR DEL REY   
5019                 6.0  75429d136f65d2d6168b9b6c5f6ec951  HINOJAR DEL REY   
5029                 5.0  75429d136f65d2d6168b9b6c5f6ec951  HINOJAR DEL REY   
7051                15.0  48000647b315f6f00f913caa757a70b3        Barcelona   
...                  ...                               ...              ...   
803888              10.0  0e49405580ff812197bf7434a92e730b        Barcelona   
837696              10.0  6acc54ea7e233502491cd124f7180d3a       Alcobendas   
837702              10.0  6acc54ea7e233502491cd124f7180d3a       Alcobendas   
887179               8.0  c71c6c26683221ceec203db2675b7df4        Alfajarín   
903926              10.0  bdea065ee5b72af9c86f0c02cc82e610       Manzanares   

       zipcode  longitud_product_id  
5010     09454                    2  
5018     09454                    2  
5019     09454                    2  
5029     09454                    2  
7051     08023                    2  
...        ...                  ...  
803888   08017                    2  
837696   28100                    2  
837702   28100                    2  
887179   50172                    2  
903926   13200                    2  

[78 rows x 12 columns]

In [17]:
df_ventas["product_id"]

0         86434
1         87652
2          2785
3         13000
4         41947
          ...  
930955     1868
930956    23539
930957    92793
930958    93036
930959    11850
Name: product_id, Length: 915165, dtype: int64

- Números de 2-5 cifras. No incluye letras, dudo que haya errores de codificación
    - Importante comprobar que estas cifras se repitan en el otro dataframe
- 5028 valores únicos. Quizás excasos para hacer regresión lineal. Se puede intentar igualmente
- Posibilidad de hacer var_obj productos comprados/no comprados (desbalanceada 25-75)
    - Es interesante la posibilidad de hacer un modelo que sea cliente que repite/no repite (100k de clientes únicos)


### Customer y num_order
De escasa utilidad de forma independiente. Propongo sustituir ambas por el número de órdenes realziadas por cada cliente.

- A continuiación se expresa ideas/código para realizar esta conversión (para regresión/productos vendidos). Esto después dehaber depurado todo lo demás.

In [18]:
# groupby("num_order").agg(moda)
# value_counts().to_dict() #Para saber el número de veces que hace pedidos cada cliente, y agruparlo como diccionario para poder sustituir)
# replace(diccionario)

- Comprobar duplicidades usando el .agg(moda) y el groupby por doble columna

### ZIPCODE

In [19]:
df_ventas["longitud_zip"] = df_ventas["zipcode"].dropna().apply(len)
df_ventas["longitud_zip"].unique().tolist()

C:\Users\Daniel\AppData\Local\Temp\ipykernel_14124\704870013.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_ventas["longitud_zip"] = df_ventas["zipcode"].dropna().apply(len)


[5.0,
 2.0,
 8.0,
 4.0,
 6.0,
 7.0,
 3.0,
 15.0,
 10.0,
 9.0,
 1.0,
 14.0,
 17.0,
 11.0,
 12.0,
 16.0,
 19.0,
 21.0,
 24.0,
 18.0,
 nan,
 22.0,
 13.0,
 20.0,
 25.0,
 29.0,
 26.0,
 27.0,
 28.0,
 23.0,
 31.0]

In [20]:
mascara_len_zip = (df_ventas["longitud_zip"] == 18)
df_ventas[mascara_len_zip]

num_order                           item_id  \
492568  2ba609d16430e3a21af898a6ab22e94a  7db9e7e793efad36394d180b227f2d3e   
492569  7c30bbd789f72279dec572705c4f5a01  87b1c2c46316245cec5d0d641af2a30a   
492570  7c30bbd789f72279dec572705c4f5a01  e88af4c6c9776b6f348bc899e054461a   
664201  f0087b3ac771471688ee465c4bd1a701  1652ea4ced4f7c2455150d63dc71ca03   
713878  3199c79301b3ab809cfdb99a2821e582  cb8304355310c971c7948df157622d38   
...                                  ...                               ...   
905651  a3236d69e4651bf5ae8415d5ea4918b1  08d07ed1ff99348d3ab53c8a6a2abdf7   
905652  a3236d69e4651bf5ae8415d5ea4918b1  12bf66f830c428c08940793a7fb02d90   
905653  b2bca6afc1f91956b0b088a0da04c242  bfa056b544f755aa565f7d642234725c   
905654  b2bca6afc1f91956b0b088a0da04c242  09b645204661c6a6bf899cf88ca068a2   
905655  b2bca6afc1f91956b0b088a0da04c242  713ec069c8b190630fee8d4a1ced85b3   

                 created_at  product_id  qty_ordered  base_cost  price  \
492568  2017-09-25 10:43:27        9611          1.0     9.0639  11.53   
492569  2018-05-29 12:05:16       38218          1.0    11.0800  14.00   
492570  2018-05-29 12:05:16        1732          1.0    15.3800  14.51   
664201  2017-10-21 13:52:38        3725          1.0    29.8000  36.90   
713878  2018-01-30 21:11:53       58199          1.0     2.0160   3.51   
...                     ...         ...          ...        ...    ...   
905651  2018-11-25 22:28:09       36269          3.0    13.0275  15.50   
905652  2018-11-25 22:28:09       61814          1.0     6.0124   7.52   
905653  2018-12-28 00:36:20        6791          3.0     5.4659   5.54   
905654  2018-12-28 00:36:20       64877          2.0     2.0228   2.97   
905655  2018-12-28 00:36:20       72360          1.0     4.1243   4.72   

        discount_percent                       customer_id       city  \
492568              15.0  3a92f6e2fbe88640bea9df526f6c9e0f    Amadora   
492569               5.0  3a92f6e2fbe88640bea9df526f6c9e0f    Amadora   
492570               5.0  3a92f6e2fbe88640bea9df526f6c9e0f    Amadora   
664201              10.0  89c65157cf230fecf2168e5927edf50c  Moscavide   
713878               7.0  ccacb015a0958aed5bb72d090f168c31      07518   
...                  ...                               ...        ...   
905651              20.0  f87983fbd9c00b7f208f2f3cbe5df885   Barcelos   
905652              20.0  f87983fbd9c00b7f208f2f3cbe5df885   Barcelos   
905653              10.0  f87983fbd9c00b7f208f2f3cbe5df885   Barcelos   
905654              10.0  f87983fbd9c00b7f208f2f3cbe5df885   Barcelos   
905655              10.0  f87983fbd9c00b7f208f2f3cbe5df885   Barcelos   

                   zipcode  longitud_product_id  longitud_zip  
492568  2610-098 Alfragide                    4          18.0  
492569  2610-098 Alfragide                    5          18.0  
492570  2610-098 Alfragide                    4          18.0  
664201  1885-100 MOSCAVIDE                    4          18.0  
713878  LLORETDEVISTALEGRE                    5          18.0  
...                    ...                  ...           ...  
905651  4750-430 Creixomil                    5          18.0  
905652  4750-430 Creixomil                    5          18.0  
905653  4750-430 Creixomil                    4          18.0  
905654  4750-430 Creixomil                    5          18.0  
905655  4750-430 Creixomil                    5          18.0  

[66 rows x 13 columns]

In [21]:
prueba_zip = df_ventas
prueba_zip["zipcode"] = prueba_zip["zipcode"].str.replace("[A-Za-z]+", "", regex=True).astype('str')

# prueba_zip = re.sub("[A-Za-z]+", "", "74ab")
# prueba_zip.astype(str)

C:\Users\Daniel\AppData\Local\Temp\ipykernel_14124\2638276310.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  prueba_zip["zipcode"] = prueba_zip["zipcode"].str.replace("[A-Za-z]+", "", regex=True).astype('str')


In [22]:
prueba_zip["zipcode"]

0         02100
1         02100
2         02100
3         02100
4         46120
          ...  
930955    23400
930956    23400
930957    08500
930958    08500
930959    08500
Name: zipcode, Length: 915165, dtype: object

In [23]:
prueba_zip["longitud_zip"] = prueba_zip["zipcode"].dropna().apply(len)
prueba_zip["longitud_zip"].unique().tolist()



C:\Users\Daniel\AppData\Local\Temp\ipykernel_14124\2243223391.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  prueba_zip["longitud_zip"] = prueba_zip["zipcode"].dropna().apply(len)


[5, 2, 8, 4, 6, 7, 3, 0, 9, 1, 11, 10, 12, 14, 13]

In [24]:
mascara_len_prueba = (prueba_zip["longitud_zip"] == 7)
prueba_zip[mascara_len_prueba]

num_order                           item_id  \
7059    0ff472fe3af657f8b22d08a9d73a8df8  3fbd6bcf73d5932c2196d80cde3755de   
7060    0ff472fe3af657f8b22d08a9d73a8df8  f42a1e771f9136d9304702e3029cdc2f   
46474   07ab22e0c5320a9c9075c18602a9354b  ea2e3b9bfbf2a084fb3320abe36aac0c   
46475   07ab22e0c5320a9c9075c18602a9354b  6e790773617d17ae8ea479bde5f2d413   
46476   07ab22e0c5320a9c9075c18602a9354b  849f1a2443059139916b07d36d922e29   
...                                  ...                               ...   
927991  08c8868547098f478b5080a68d33113d  e01399ff73f93aba3ceba98056a3571b   
927992  08c8868547098f478b5080a68d33113d  6996f4ce991b0c6192f6bb2a3e00d51c   
928194  544c5bcda59319fe1eb4195d469f0e63  44d08fec529ac49c77dfa762f92a28fd   
928195  544c5bcda59319fe1eb4195d469f0e63  e7b2159c8cc976adedc3eeae029390b6   
928242  0e89db0f9c966440f00ab89c93d5de9a  d6f43cbb6d1b5b8629bac2f34c2879af   

                 created_at  product_id  qty_ordered  base_cost  price  \
7059    2017-03-13 12:09:47       58940          1.0    18.6600  25.87   
7060    2017-03-13 12:09:47       40513          1.0     8.4720  11.50   
46474   2017-03-20 13:32:23       39469          1.0    16.4513  19.79   
46475   2017-03-20 13:32:23       12156          1.0    10.9855  12.30   
46476   2017-03-20 13:32:23       33925          1.0    24.0000  37.49   
...                     ...         ...          ...        ...    ...   
927991  2018-12-25 18:08:23       67864          1.0     1.2325   1.65   
927992  2018-12-25 18:08:23       65129          1.0     5.7000   6.57   
928194  2018-12-26 09:17:34        7635          1.0    23.9330  29.44   
928195  2018-12-26 09:17:34        7195          1.0     6.7645   8.12   
928242  2018-12-26 11:46:43       59393          2.0     3.1949   3.83   

        discount_percent                       customer_id               city  \
7059                 5.0  61d009da208a34ae155420e55f97abc7  Santanyí/Son Moja   
7060                 5.0  61d009da208a34ae155420e55f97abc7  Santanyí/Son Moja   
46474                5.0  ed0eebbaa23beae2d9fcb2fd70cd6feb            LIGARES   
46475                5.0  ed0eebbaa23beae2d9fcb2fd70cd6feb            LIGARES   
46476                5.0  ed0eebbaa23beae2d9fcb2fd70cd6feb            LIGARES   
...                  ...                               ...                ...   
927991              10.0  18c68d2f9e79e5e73d4fe64588ac9434     porto Gondomar   
927992              10.0  18c68d2f9e79e5e73d4fe64588ac9434     porto Gondomar   
928194              10.0  ecd4746d80847f5e13a60ff191ac3f8d             Loures   
928195              10.0  ecd4746d80847f5e13a60ff191ac3f8d             Loures   
928242              50.0  b2947b307a98eef9f3078823a543d8ab            Logroño   

        zipcode  longitud_product_id  longitud_zip  
7059     07650                     5             7  
7060     07650                     5             7  
46474   5180301                    5             7  
46475   5180301                    5             7  
46476   5180301                    5             7  
...         ...                  ...           ...  
927991  4420126                    5             7  
927992  4420126                    5             7  
928194  2680273                    4             7  
928195  2680273                    4             7  
928242  26003 (                    5             7  

[596 rows x 13 columns]

- Mucho trabajo pro delante
- Lo primero sustituir letras por ""
- Utilizar un webscrapping mas avanzado
- Utilizar recursivamente ambas columnas para corregir posibles problemas de una y otra
- Sustituir valores "" por NaN

### Valores perdidos

In [25]:
mascara_na = df_ventas["zipcode"].isna()

In [28]:
df_ventas


num_order                           item_id  \
0       ce30c2f02458457e3c7b563a636ae2a1  0916c05c5c3f65f59d813a78ac35c8d2   
1       ce30c2f02458457e3c7b563a636ae2a1  ff323b39ae36843396d2e53ce549fb10   
2       ce30c2f02458457e3c7b563a636ae2a1  199916dffc95259f4d2daab6664ca9c0   
3       83e75d608f11c8163599806420903ab9  8ca334ec2493501139327ce0165a1a84   
4       9b687270d8e7eed9717022af5961a190  ce0a1683c6b1a0248b330344ec592ddf   
...                                  ...                               ...   
930955  479ed436c26a511c4c3356e679bccbc9  37d700d54a6acc0ff49392cce34b6427   
930956  479ed436c26a511c4c3356e679bccbc9  05fd04261f1c951b07f2ac3a67254d81   
930957  bf923aa16e0608a5ed95347a9f411dce  8d8600ea4e451330f8614ba08ad4f047   
930958  bf923aa16e0608a5ed95347a9f411dce  b74b2e1a096c656a247142e24eefb738   
930959  bf923aa16e0608a5ed95347a9f411dce  fd0c1bb9a90dce3f3b2099990f019262   

                 created_at  product_id  qty_ordered  base_cost  price  \
0       2018-11-06 16:52:13       86434          1.0     9.3750  15.38   
1       2018-11-06 16:52:13       87652          1.0     7.0000  15.41   
2       2018-11-06 16:52:13        2785          1.0     9.1587  12.23   
3       2018-12-17 12:26:54       13000          5.0    13.7936  15.03   
4       2017-01-12 14:19:03       41947          1.0     5.5924   7.00   
...                     ...         ...          ...        ...    ...   
930955  2018-12-30 23:51:57        1868          1.0     4.0700   4.79   
930956  2018-12-30 23:51:57       23539          1.0     4.0700   5.35   
930957  2018-12-30 23:57:37       92793          1.0    33.7360  32.93   
930958  2018-12-30 23:57:37       93036          1.0     0.0000  16.32   
930959  2018-12-30 23:57:37       11850          1.0    10.4414  12.53   

        discount_percent                       customer_id  \
0                    8.0  eccbc87e4b5ce2fe28308fd9f2a7baf3   
1                    8.0  eccbc87e4b5ce2fe28308fd9f2a7baf3   
2                    8.0  eccbc87e4b5ce2fe28308fd9f2a7baf3   
3                   50.0  eccbc87e4b5ce2fe28308fd9f2a7baf3   
4                    5.0  1679091c5a880faf6fb5e6087eb1b2dc   
...                  ...                               ...   
930955              10.0  d9cd4886b615a36059f72a74d7019113   
930956              10.0  d9cd4886b615a36059f72a74d7019113   
930957              10.0  d73f27d3ab96df71f0fdc47511f326bc   
930958              10.0  d73f27d3ab96df71f0fdc47511f326bc   
930959              10.0  d73f27d3ab96df71f0fdc47511f326bc   

                         city zipcode  longitud_product_id  longitud_zip  
0       TARAZONA DE LA MANCHA   02100                    5             5  
1       TARAZONA DE LA MANCHA   02100                    5             5  
2       TARAZONA DE LA MANCHA   02100                    4             5  
3       TARAZONA DE LA MANCHA   02100                    5             5  
4                    Alboraya   46120                    5             5  
...                       ...     ...                  ...           ...  
930955                  Ubeda   23400                    4             5  
930956                  Ubeda   23400                    5             5  
930957                    Vic   08500                    5             5  
930958                    Vic   08500                    5             5  
930959                    Vic   08500                    5             5  

[915165 rows x 13 columns]

In [27]:
df_ventas[mascara_na].isna().sum()

num_order              0.0
item_id                0.0
created_at             0.0
product_id             0.0
qty_ordered            0.0
base_cost              0.0
price                  0.0
discount_percent       0.0
customer_id            0.0
city                   0.0
zipcode                0.0
longitud_product_id    0.0
longitud_zip           0.0
dtype: float64